README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

Universal Web Image for Flutter Web apps with external hosted images.

## Features

* Show an image as html image, if necessary.
* Distinguish between svg images and others automatically

## Motivation

If you try to load an image without valid XHR status 200, an error is thrown and the image does not
become visible on web.

```
Failed to load network image.
Image URL: https://example.com/invalidConfiguredServerImage.jpg
Trying to load an image from another domain? Find answers at:
https://flutter.dev/docs/development/platform-integration/web-images
```

This package solves the issue by using a Web Image as fallback.

## Usage

```dart
import 'package:flutter/material.dart';
import 'package:universal_web_image/universal_web_image.dart';

class MyHomePage extends StatelessWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  // 424x283
  static const validImageUrl = 'https://www.kesselheld.de/content/uploads/2018/09/durchschnittliche-heizkosten.jpg';

  // 640x300
  static const invalidImageUrl = 'https://img.utopia.de/cH68CrxH2UEqUr8eVAWC-BCgXoU=/640x300/https://utopia.de/app/uploads/2019/02/richtig-heizen-wpb-ivan-kruk-ri-190820-1280x720.jpg';

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.start,
          children: [
            const Text(
              'Web Image with valid XHR 200 status',
            ),
            const UniversalWebImage(
              imageUrl: validImageUrl,
              width: 424,
              height: 283,
            ),
            const Text(
              'Web Image with invalid XHR 0 status',
            ),
            const UniversalWebImage(
              imageUrl: invalidImageUrl,
              width: 640,
              height: 300,
            ),
            const Text(
              'Regular image with invalid XHR 0 status',
            ),
            Image.network(
              invalidImageUrl,
              width: 640,
              height: 300,
            ),
          ],
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
```


CHANGELOG.md
## 0.1.0

* Be able to attach a custom loader widget
* Optional onLoaded callback added

## 0.0.3

BoxFit fix

## 0.0.2

Do not apply width/height if BoxFit is set

## 0.0.1

Inital release

test/universal_web_image_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:universal_web_image/universal_web_image.dart';

void main() {
  test('test', () {});
}


lib/universal_web_image.dart
library universal_web_image;

import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';

import 'package:http/http.dart' as http;
import 'package:universal_web_image/web_image/canvas_renderer_image.dart';

import 'package:universal_web_image/web_image/web_image.dart'
    if (dart.library.js) 'package:universal_web_image/web_image/web_image_web.dart';

/// If there are loading problems with an image on the web, then this view
/// automatically uses HTML to render the image, otherwise an image is
/// drawn on the flutter canvas.
class UniversalWebImage extends StatefulWidget {
  const UniversalWebImage({
    super.key,
    required this.imageUrl,
    this.width,
    this.height,
    this.fit,
    this.customImageBuilder,
    this.customLoaderBuilder,
  });

  final String imageUrl;

  final BoxFit? fit;

  final double? width;

  final double? height;

  /// If you use [customImageBuilder], you break the auto distinguish feature.
  /// Hence, you have to handle drawing svg images vs. other images yourself.
  final Widget Function(BuildContext context)? customImageBuilder;

  /// Use to give custom loading widget while image is loading
  final Widget Function(BuildContext context)? customLoaderBuilder;

  @override
  State<UniversalWebImage> createState() => _UniversalWebImageState();
}

class _UniversalWebImageState extends State<UniversalWebImage> {
  Uint8List? bytes;

  bool hasError = false;

  @override
  Widget build(BuildContext context) {
    if (kIsWeb && bytes == null && !hasError) {
      _bytes();
      return widget.customLoaderBuilder?.call(context) ?? const Text('loading');
    }
    if (!hasError && widget.customImageBuilder == null) {
      return CanvasRendererImage(
          imageUrl: widget.imageUrl,
          width: widget.width,
          height: widget.height,
          fit: widget.fit,
          onLoaded: () {
            setState(() {});
          });
    }
    if (!hasError && widget.customImageBuilder != null) {
      return widget.customImageBuilder!(context);
    }

    // Image ${widget.imageUrl} cannot be drawn by Flutter canvas renderer.
    return webImage(
      context: context,
      imageUrl: widget.imageUrl,
      // TODO: How can be get pixels of the image?
      width: widget.width ?? 100,
      height: widget.height ?? 100,
    );
  }

  Future<bool> _bytes() async {
    try {
      final response = await http.get(Uri.parse(widget.imageUrl));
      setState(() {
        bytes = response.bodyBytes;
      });
      return true;
    } catch (e) {
      if (e is http.ClientException && e.message == 'XMLHttpRequest error.') {
        setState(() {
          hasError = true;
        });
        return false;
      } else {
        debugPrint(e.toString());
      }
    }
    return true;
  }
}


lib/web_image/canvas_renderer_image.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_svg/flutter_svg.dart';

/// This image widget uses the Flutter rendering engine (drawing on the canvas).
class CanvasRendererImage extends StatelessWidget {
  const CanvasRendererImage({
    super.key,
    required this.imageUrl,
    this.bytes,
    this.width,
    this.height,
    this.fit,
    this.onLoaded,
  });

  final String imageUrl;

  final Uint8List? bytes;

  final double? width;

  final double? height;

  final BoxFit? fit;

  final VoidCallback? onLoaded;

  @override
  Widget build(BuildContext context) {
    final bool isSvgImage = imageUrl.endsWith('.svg');

    if (bytes != null) {
      return isSvgImage
          ? SvgPicture.memory(
              bytes!,
              height: height,
              width: width,
              fit: fit ?? BoxFit.contain,
            )
          : Image.memory(
              bytes!,
              height: height,
              width: width,
              fit: fit,
            );
    }
    return isSvgImage
        ? SvgPicture.network(
            imageUrl,
            height: height,
            width: width,
            fit: fit ?? BoxFit.contain,
          )
        : Image.network(
            imageUrl,
            width: width,
            height: height,
            fit: fit,
            loadingBuilder: (context, widget, event) {
              if (event != null &&
                  event.cumulativeBytesLoaded == event.expectedTotalBytes &&
                  onLoaded != null) {
                onLoaded!();
              }
              return widget;
            },
          );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:universal_web_image/universal_web_image.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Universal Web Image Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Universal Web Image Demo Page'),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  // 424x283
  static const validImageUrl =
      'https://www.kesselheld.de/content/uploads/2018/09/durchschnittliche-heizkosten.jpg';

  // 640x300
  static const invalidImageUrl =
      'https://img.utopia.de/cH68CrxH2UEqUr8eVAWC-BCgXoU=/640x300/https://utopia.de/app/uploads/2019/02/richtig-heizen-wpb-ivan-kruk-ri-190820-1280x720.jpg';

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.start,
          children: [
            const Text(
              'Web Image with valid XHR 200 status with custom loader',
            ),
            UniversalWebImage(
              imageUrl: validImageUrl,
              customLoaderBuilder: (context) =>
                  const Center(child: CircularProgressIndicator()),
            ),
            const Text(
              'Web Image with invalid XHR 0 status',
            ),
            const UniversalWebImage(
              imageUrl: invalidImageUrl,
              width: 640,
              height: 300,
            ),
            const Text(
              'Regular image with invalid XHR 0 status',
            ),
            Image.network(
              invalidImageUrl,
              width: 640,
              height: 300,
            ),
          ],
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
